### DISCUSSION TITLE
Problems with circular references

### DISCUSSION THREAD

**Discussion by andresmanz** [mswjs/data#208] [Ideas]
Hi,

I stumbled across a circular reference problem with this db:

```typescript
const db = factory({
  user: {
    id: primaryKey(Number),
    name: String,
    memberships: manyOf('membership'),
  },
  team: {
    id: primaryKey(Number),
    name: String,
    memberships: manyOf('membership'),
  },
  membership: {
    id: primaryKey(Number),
    role: String,
    user: oneOf('user'),
    team: oneOf('team'),
  },
});

const user = db.user.create({ id: 1, name: 'Test User' });
const team = db.team.create({ id: 1, name: 'Test Team' });
const membership = db.membership.create({
  id: 1,
  role: 'admin',
  user: user,
  team: team,
});

// creating the membership does *not* add it to the user's memberships. an explicit update is required:
const updatedUser = db.user.update({
  where: { id: { equals: 1 } },
  data: {
    memberships: [membership],
  },
});
```

So in this example I create a user and a team and then add the user to the team as an admin.
In the test for a member list, two endpoints are called:
- `/users/me` returns the user with their memberships
- `/teams/:id/memberships` returns a team's memberships

However, I noticed that MSW's response for `/users/me` was always null.
JSON.stringify(updatedUser) throws the following error because the user and the membership reference each other:

```
TypeError: Converting circular structure to JSON
        --> starting at object with constructor 'Object'
        |     property 'memberships' -> object with constructor 'Array'
        |     index 0 -> object with constructor 'Object'
        --- property 'user' closes the circle
        at JSON.stringify (<anonymous>)
```

This issue could be worked around, but it could get annoying quickly. So I'd like to provide some feedback, mostly based on ORM experience:
1. When creating a membership with a user, the membership should also automatically be added to the user's memberships. Defining the inverse of the relationship might be necessary for this to work. Not sure if this is related to #92.
2. There should be a way to explicitly populate relationships when loading entities. I guess this implies returning
related objects as copies instead of references.
3. It would be nice to link foreign key columns to relations. I'll provide an example below.

Nr 2 would solve the circular reference problem. Nr 1 could potentially make testing a bit easier and more intuitive. I'd imagine something like the following code to work:

```typescript
const db = factory({
  user: {
    id: primaryKey(Number),
    name: String,
    memberships: manyOf('membership', membership => membership.user), // <-- specify the inverse
  },
  team: {
    id: primaryKey(Number),
    name: String,
    memberships: manyOf('membership', membership => membership.team),
  },
  membership: {
    id: primaryKey(Number),
    role: String,
    user: oneOf('user', user => user.memberships),
    team: oneOf('team', team => team.memberships),
  },
});

const user = db.user.create({ id: 1, name: 'Test User' });
const team = db.team.create({ id: 1, name: 'Test Team' });
const membership = db.membership.create({
  id: 1,
  role: 'admin',
  user: user,
  team: team,
});

// Feedback #1: no need to add the membership to user explicitly, reloading is enough (should be possible by defining inverse relationship)
const updatedUser = db.user.findFirst({
  where: { id: { equals: 1 } },
  // Feedback #2: ability to populate
  populate: ['memberships', 'memberships.team']
});
```

expected result:
```
{
  id: 1,
  name: 'Test User',
  memberships: [
      {
          id: 1,
          role: 'admin',
          // user: undefined or null or user id
          team: {
            id: 1,
            name: 'Test Team',
          }
      }
  ]
}
```

Nr 3, as a bonus: Sometimes you need to access a related object's id without populating the related object. You can set the id explicitly along with the related object, but it would be nice to have msw data do this automatically. Example:

```typescript
const db = factory({
  ...
  membership: {
    id: primaryKey(Number),
    role: String,
    userId: Number, // will be set/updated automatically whenever user entity is set
    user: oneOf('user', { foreignKey: 'userId' }),
    teamId: Number,
    team: oneOf('team', { foreignKey: 'teamId' }),
  },
});
```

I'm sorry if this is stupid or doesn't align with the plans for msw data at all. Feel free to ignore in that case. Also, sorry for that huge wall of text :(

**Comment by kettanaito** [comment]
Hey, @andresmanz. Thank you for raising this!

The Data library does not support inverse relationships in the sense that you expect them to work. I'd very much like to add the support for that though! 

Not sure what I think about the `populate` example so far. We should start it small and support inverse the barebones of inverse relationships and then explore additional scenarios on top. Perhaps that will give me more context and I'd be able to judge your suggestions better. 

We can start by writing the logic that'd allow the library to infer inverse relationships based on model definitions.
